storage/pebble/procedure/children.go

```diff
@@ -24,8 +24,10 @@ import (
 //     there are two special cases for (2):
 //     - if the parent block is zero, then we don't need to add this index.
 //     - if the parent block doesn't exist, then we will insert the child index instead of updating
-func IndexNewBlock(blockID flow.Identifier, parentID flow.Identifier) func(*badger.Txn) error {
-	return func(tx *badger.Txn) error {
+func IndexNewBlock(blockID flow.Identifier, parentID flow.Identifier) func(storage.PebbleReaderBatchWriter) error {
+	return func(rw storage.PebbleReaderBatchWriter) error {
+		r, tx := rw.ReaderWriter()
+
 		// Step 1: index the child for the new block.
 		// the new block has no child, so adding an empty child index for it
 		err := operation.InsertBlockChildren(blockID, nil)(tx)
```

Comment:
## ⚠️ 

This has now become intrinsically unsafe and prone to breaking the protocol state's consistency. I think we need a larger refactoring here, where we move the logic to inserting blocks and indexing children to one specific component that has the sole authority to maintain these indices. 

Previously this code was intrinsically safe: it would fail if the block was already present before (possibly already having children indexed). Now we cannot _reliably_ check anymore that this block isn't in the database yet. If we ever index a block that was already indexed before, we will _overwrite its children with nil_. I think that is an unacceptable surface for accidentally breaking safety-critical assumptions in the future.